t/lib/CorpusTest.pm

```diff
@@ -64,56 +104,150 @@ sub _validity_tests {
-        my $lossy = $case->{lossy};
-
-        my $B = $case->{bson};
-        my $E = $case->{extjson}; # could be undef
-
-        my $cB = exists($case->{canonical_bson}) ? $case->{canonical_bson} : $B;
-        my $cE = exists($case->{canonical_extjson}) ? $case->{canonical_extjson} : $E;
-
-        my $skip_extjson = !(defined($E) && _extjson_ok($bson_type, $E));
-
-        $B = pack( "H*", $B );
-        $cB = pack( "H*", $cB );
-
-        $E = _normalize( $E, "$desc: normalizing E"  );
-        $cE = _normalize( $cE, "$desc: normalizing cE"  );
-
-        _bson_to_bson( $codec, $B, $cB, "$desc: B->cB" );
-
-        if ($B ne $cB) {
-            _bson_to_bson( $codec, $cB, $cB, "$desc: cB->cB" );
-        }
+        subtest 'case: '.$case->{description} => sub {
+            local $Data::Dumper::Useqq = 1;
+
+            my $desc = $case->{description};
+            ok 1, 'noop'; # ensure there's a success for empty groups
+            my $wrap = $bson_type =~ /\A(?:0x00|0x01|0x10|0x12)\z/;
+            my $codec = BSON->new( prefer_numeric => 1, wrap_numbers => $wrap, ordered => 1 );
+            my $lossy = $case->{lossy};
+
+            my $canonical_bson = $case->{canonical_bson} || $case->{bson};
+            my $converted_bson = $case->{converted_bson};
+            my $degenerate_bson = $case->{degenerate_bson};
+
+            $canonical_bson = pack('H*', $canonical_bson);
+            $converted_bson = pack('H*', $converted_bson)
+                if defined $converted_bson;
+            $degenerate_bson = pack('H*', $degenerate_bson)
+                if defined $degenerate_bson;
+
+            my $canonical_json = $case->{canonical_extjson};
+            my $converted_json = $case->{converted_extjson};
+            my $degenerate_json = $case->{degenerate_extjson};
+            my $relaxed_json = $case->{relaxed_extjson};
+
+            $canonical_json = _normalize(
+                $canonical_json,
+                '$desc: normalizing canonical extjson',
+            );
+            $converted_json = _normalize(
+                $converted_json,
+                '$desc: normalizing converted extjson',
+            );
+            $degenerate_json = _normalize(
+                $degenerate_json,
+                '$desc: normalizing degenerate extjson',
+            );
+            $relaxed_json = _normalize(
+                $relaxed_json,
+                '$desc: normalizing relaxed extjson',
+            );
+
+            my $has_canonical_bson = defined $canonical_bson;
+            my $has_converted_bson = defined $converted_bson;
+            my $has_degenerate_bson = defined $degenerate_bson;
+
+            my $has_canonical_json = defined $canonical_json;
+            my $has_converted_json = defined $converted_json;
+            my $has_relaxed_json = defined $relaxed_json;
+            my $has_degenerate_json = defined $degenerate_json;
+
+            if (!$deprecated and $has_canonical_bson and $has_canonical_json) {
```

Comment:
I find it hard to follow all the conditionals and map them back to the spec assertions.  Could you please refactor them at least into two functions: one for deprecated and one for non-deprecated.

Then I think some of the conditionals are unnecessary as they are certain to exist (e.g. canonical_bson and canonical_json).